### ISSUE TITLE
Emacs 29 issue on Windows

### ISSUE THREAD

**Issue by cuprum** [ahyatt/ekg#57]
I have been trying Emacs 29 recently and found out that when I created a note in Emacs 29, I can't view it in Emacs 28 (with `ekg-show-notes-with-tag`, for example) and vice versa.
This is what the notes look like in the SQLbrowser when created in different versions of Emacs. Please, note the right column:

![28](https://github.com/ahyatt/ekg/assets/532546/8fb0e4e7-64f1-4708-ba35-1314ce835579)

![29](https://github.com/ahyatt/ekg/assets/532546/6daa02eb-3286-4c2d-abda-7d1a049a3c45)

Perhaps, built-in support for SQLite in 29th version has an effect here?

Info: develop branch of EKG, GNU Emacs 29.0.90 (build 3, x86_64-w64-mingw32) and GNU Emacs 28.1 (build 2, x86_64-w64-mingw32)

**Comment by cuprum** [comment]
@ahyatt Sorry, I was late, there was no opportunity to be at the computer.

>If you update the properties column to be () instead of (:t t), does it solve your problems?

No, unfortunately.

**Comment by ahyatt** [comment]
Interesting.  I've been unable to reproduce a problem just on the triples package at the moment.  I'll try to reproduce it in ekg.  I've previously done this just to make sure there is compatibility, but let me try again.

**Comment by ahyatt** [comment]
I've checked in a change [Fix emacsql/buit-in sqlite incompatibility](https://github.com/ahyatt/triples/commit/45dc57703ae870aa864a195041b76d622535c351), to fix the underlying cause.  However, we also need an upgrade mechanism for the built-in sqlite users.  But first, please let me know if using the newest `develop` branch of the `triples` package fixes the issue for you, allowing you to see your old data.

**Comment by cuprum** [comment]
With develop branch of the Triples new notes as well as single old notes are displayed in both versions of Emacs.

But I have a borderline case where one note was created in Emacs 28 and the other in Emacs 29 under same tag (using Triple 0.27). Unfortunately, when using the develop branch of the triples, the error still occurs. For example, when I call `ekg-show-notes-with-tag` in Emacs 28, backtrace is:
```
Debugger entered--Lisp error: (wrong-type-argument number-or-marker-p nil)
  <(nil 1685348741)
  ekg-sort-by-creation-time(#s(ekg-note :id 33665031154 :text nil :mode nil :tags nil :creation-time nil :modified-time nil :properties nil :inlines nil) #s(ekg-note :id 33665031782 :text "tag1 e28" :mode org-mode :tags ("date/2023-05-29" "tag1") :creation-time 1685348741 :modified-time 1685348772 :properties nil :inlines nil))
  sort((#s(ekg-note :id 33665031782 :text "tag1 e28" :mode org-mode :tags ("date/2023-05-29" "tag1") :creation-time 1685348741 :modified-time 1685348772 :properties nil :inlines nil)) ekg-sort-by-creation-time)
  (closure ((tag . "tag1") ekg-notes-mode-abbrev-table ekg-notes-mode-syntax-table cl-struct-ekg-inline-tags cl-struct-ekg-note-tags t) nil (sort (ekg-get-notes-with-tag tag) #'ekg-sort-by-creation-time))()
  funcall((closure ((tag . "tag1") ekg-notes-mode-abbrev-table ekg-notes-mode-syntax-table cl-struct-ekg-inline-tags cl-struct-ekg-note-tags t) nil (sort (ekg-get-notes-with-tag tag) #'ekg-sort-by-creation-time)))
  (mapc #'(lambda (note) (ewoc-enter-last ewoc note)) (funcall notes-func))
  (let ((ewoc (ewoc-create #'ekg-display-note-insert (propertize name 'face 'ekg-notes-mode-title)))) (mapc #'(lambda (note) (ewoc-enter-last ewoc note)) (funcall notes-func)) (ekg-notes-mode) (progn (set (make-local-variable 'ekg-notes-ewoc) ewoc) (set (make-local-variable 'ekg-notes-fetch-notes-function) notes-func) (set (make-local-variable 'ekg-notes-name) name) (set (make-local-variable 'ekg-notes-hl) (make-overlay 1 1)) (set (make-local-variable 'ekg-notes-tags) tags)) (overlay-put ekg-notes-hl 'face hl-line-face) (forward-line 1) (ekg--note-highlight))
  ekg--show-notes(#("tag: tag1" 5 9 (face ekg-tag)) (closure ((tag . "tag1") ekg-notes-mode-abbrev-table ekg-notes-mode-syntax-table cl-struct-ekg-inline-tags cl-struct-ekg-note-tags t) nil (sort (ekg-get-notes-with-tag tag) #'ekg-sort-by-creation-time)) ("tag1"))
  (let ((buf (get-buffer-create (format "*ekg %s*" name)))) (set-buffer buf) (ekg--show-notes name notes-func tags) (pop-to-buffer buf))
  ekg-setup-notes-buffer(#("tag: tag1" 5 9 (face ekg-tag)) (closure ((tag . "tag1") ekg-notes-mode-abbrev-table ekg-notes-mode-syntax-table cl-struct-ekg-inline-tags cl-struct-ekg-note-tags t) nil (sort (ekg-get-notes-with-tag tag) #'ekg-sort-by-creation-time)) ("tag1"))
  ekg-show-notes-with-tag("tag1")
  funcall-interactively(ekg-show-notes-with-tag "tag1")
  command-execute(ekg-show-notes-with-tag record)
  execute-extended-command(nil "ekg-show-notes-with-tag" nil)
  funcall-interactively(execute-extended-command nil "ekg-show-notes-with-tag" nil)
  command-execute(execute-extended-command)
```

**Comment by jayrajput** [comment]
Did you upgraded the triples db to make sure all the subject id as integers? Something like this 

```
(triples-upgrade-to-0.3 ekg-db)
```

I was also getting the same sort error.

**Comment by ahyatt** [comment]
I've now checked in changes to the `develop` branch so that it should automatically upgrade the next time you use ekg.  Please let me know if this does or does not solve the problem.  Otherwise, I believe I've solved the problem, so will close this soon.

**Comment by cuprum** [comment]
@ahyatt I'm afraid this doesn't solve the problem.
Here is a database with 2 notes created under the following conditions: Emacs 28, Ekg 3.0 (main branch), Triples 0.27. One note was created in Emacs 28, the other in Emacs 29.
[triples.zip](https://github.com/ahyatt/ekg/files/11646561/triples.zip)
Then after updating to develop branches Ekg and Triples and calling `ekg-show-notes-with-tag` "tag" the database is
[triples-after.zip](https://github.com/ahyatt/ekg/files/11646565/triples-after.zip)
And the error message still appears.

**Comment by ahyatt** [comment]
Thank you for the examples!  I took a look and couldn't reproduce the issue in the first one - the upgrade in the latest `develop` branch seems to have worked.  In the second, I could reproduce the issue.  Looking at the file, it looks like it was marked as upgraded via the new code, but the integer strings were not converted to integers like they should have been.

One way this might happen is if you were running emacs 29 and upgrading with emacsql instead of the emacs 29 built-in sqlite (specifically what happened is that `triples-sqlite-interface` was `emacsql` instead of `builtin`).   Do you think this could be what happened?  Otherwise I'll have to think of some other ways that you could have gone through `ekg-upgrade-db` without actually changing the string integers.  

If this is true, perhaps I should force the triples package to do the upgrade.

**Comment by cuprum** [comment]
@ahyatt This seems to be my mistake. I still use emacs 28 as the main version and emacs 29 for tests. So, after upgrading to the to develop branches of Ekg and Triples I start Emacs 28,  call `ekg-show-notes-with-tag` and an error appears. After that the error will also appear in Emacs 29.

**Comment by ahyatt** [comment]
FYI I've completely changed how we upgrade in the latest `triples` code.  I noticed, in my own repository, some corruption, so I traced it down to more issues with incompatibilities plus how we were fixing everything. I restored it from a backup and ran the latest code, and it looks much better now.  You may want to check it out. For users who started out with the builtin sqlite, only the latest code will truly fix the sqlite / emacsql incompatibilities.  For users who started out with emacsql, the latest upgrade code is just a bit safer, and I suspect if you just are concerned about a few entries since upgrading to emacs 29, it won't make a big difference.

**Comment by ahyatt** [comment]
Another note about this - you may want to check the state of your database; both my database I use, when upgraded, had no indices.  You can tell by whether the following outputs anything or not:

```
SELECT name, sql
FROM sqlite_master
WHERE type='index' AND tbl_name='triples';
```

If there's an issue, the new upgrade code in the `triples` packages fixes it, and I've added `ekg-force-upgrade` to the latest commit on the `develop` branch, so you can run it and then check again.  It should fix it; it worked for both of my test databases.

**Comment by cuprum** [comment]
Thank you very much! I have already upgraded with previous commit. In Emacs 29 everything is great. But, in Emacs 28 when i try to call `ekg-show-notes-with-tag`  i get an error:
```
Debugger entered--Lisp error: (file-missing "Cannot open load file" "No such file or directory" "sqlite")
  require(sqlite)
  load-with-code-conversion("c:/Users/Deb/AppData/Roaming/.emacs.d/elisp/triples/triples-upgrade.el" "c:/Users/Deb/AppData/Roaming/.emacs.d/elisp/triples/triples-upgrade.el" nil t)
  require(triples-upgrade)
  load-with-code-conversion("c:/Users/Deb/AppData/Roaming/.emacs.d/elisp/ekg/ekg.el" "c:/Users/Deb/AppData/Roaming/.emacs.d/elisp/ekg/ekg.el" nil t)
  command-execute(ekg-show-notes-with-tag record)
  execute-extended-command(nil "ekg-show-notes-with-tag" nil)
  funcall-interactively(execute-extended-command nil "ekg-show-notes-with-tag" nil)
  command-execute(execute-extended-command)
```

**Comment by ahyatt** [comment]
Thanks - I fixed the problem in the latest version of the triples library, I'll upgrade the ekg dependency when it goes live.

**Comment by ahyatt** [comment]
That's great, thank you!  I'll close this, then, and release ekg 0.3.1 within a few days.
